### PR TITLE
Do not allow memory regions larger than (1ull << MAX_PADDR_BITS)

### PR DESCRIPTION
**Pull request by aap-sc**
Such regions are inaccessible to the hardware being modeled.

### CODE REVIEW

**Comment by aap-sc** [comment]
This commit addresses an issue  (discussed in #1100) when user can configure memory region to have addresses larger than (1 << MAX_PADDR_BITS), while the current spike implementation does not really allow for such accesses.

File: spike_main/spike.cc [hunk 1]
```diff
@@ -189,7 +189,20 @@ static std::vector<mem_cfg_t> parse_mem_layout(const char* arg)
       exit(EXIT_FAILURE);
     }
 
-    res.push_back(mem_cfg_t(reg_t(base), reg_t(size)));
+    mem_cfg_t mem_region(base, size);
+    const uint64_t max_allowed_pa = (1ull << MAX_PADDR_BITS) - 1ull;
+    if ((mem_region.base + mem_region.size - 1) > max_allowed_pa) {
+      fprintf(stderr, "unsupported memory region "
+                      "{ base = 0x%llX, size = 0x%llX } specified\n"
```
Comment (scottj97) [inline spike_main/spike.cc]: [Use `cinttypes`](https://stackoverflow.com/a/9225648/202201) for printing `uint64_t`: `PRIX64`

Or C++ streams to `cerr`.

File: spike_main/spike.cc [hunk 2]
```diff
@@ -189,7 +189,20 @@ static std::vector<mem_cfg_t> parse_mem_layout(const char* arg)
...
+                      "currently, spike does not support physical adresses "
```
Comment (scottj97) [inline spike_main/spike.cc]: addresses is misspelled

File: spike_main/spike.cc [hunk 3]
```diff
@@ -189,7 +189,20 @@ static std::vector<mem_cfg_t> parse_mem_layout(const char* arg)
...
+                      "larger than 0x%llX)\n",
```
Comment (scottj97) [inline spike_main/spike.cc]: closing parenthesis has no matching open parenthesis

File: spike_main/spike.cc (hunk 2 above)
Comment (scottj97) [inline spike_main/spike.cc]: How about we give the user a clue within the error message about why the physical address is limited?

"Based on MAX_PADDR_BITS, the largest accessible physical address is 0x%x"

**Review by scottj97 — CHANGES_REQUESTED** [review]
I'd like the commit message to explain better why this change is good. It's not that "currently, spike may not support such memory regions properly" it's that "Such regions are inaccessible to the hardware being modeled."

**Comment by liweiwei90** [comment]
I just have another question. Do we needs to take the other device memory map into consideration? For example,

- If we use the memory cfg: 0:0x100000000, if we try to access 0x02000004(in both the address space for memory and clint), it will access clint (upper_bound seems take the base address to compare, any misunderstanding?)

- And, if we try to access 0x02100000(out of the clint address space, but in memory address space), however, it seems also will access clint and fail.

File: spike_main/spike.cc (hunk 1 above)
Comment (aap-sc) [reply spike_main/spike.cc]: Thanks! Decided to use PRIX64 since we tend to use fprintf in other places.

File: spike_main/spike.cc (hunk 2 above)
Comment (aap-sc) [reply spike_main/spike.cc]: my bad, fixed.

File: spike_main/spike.cc (hunk 3 above)
Comment (aap-sc) [reply spike_main/spike.cc]: fixed

File: spike_main/spike.cc (hunk 2 above)
Comment (aap-sc) [reply spike_main/spike.cc]: Sure. Updated the error message.

**Comment by aap-sc** [comment]
> I just have another question. Do we needs to take the other device memory map into consideration?

I guess that we should, though I don't think that the current spike interface allows that. To add a device to the bus we specify only the base address. The size of the underlying memory region is not specified.

> And, if we try to access 0x02100000(out of the clint address space, but in memory address space), however, it seems also will access clint and fail.

I guess this is what will happen. 

It is also may be the case that your question should be treated a different issue - that is it does not look like it is **directly** related to this PR.

File: spike_main/spike.cc (hunk 1 above)
Comment (aap-sc) [reply spike_main/spike.cc]: One thing to note - to avoid compiler warnings I had to update the surrounding code a little bit (auto -> explicit uint64_t and started to use PRIx64 in another error message). Hope this is not  a big deal.

**Comment by liweiwei90** [comment]
> It is also may be the case that your question should be treated a different issue - that is it does not look like it is **directly** related to this PR.

Yeah. The current PR is LGTM.  It's just another question that may also be used to limit the memory region.

**Comment by aap-sc** [comment]
@scottj97 , gentle ping. Could you please take a look once again?

File: spike_main/spike.cc [hunk 4]
```diff
@@ -176,20 +177,32 @@ static std::vector<mem_cfg_t> parse_mem_layout(const char* arg)
       size += PGSIZE - size % PGSIZE;
 
     if (size != size0) {
-      fprintf(stderr, "Warning: the memory at  [0x%llX, 0x%llX] has been realigned\n"
-                      "to the %ld KiB page size: [0x%llX, 0x%llX]\n",
+      fprintf(stderr, "Warning: the memory at "
+                      "[0x%" PRIx64 ", 0x%" PRIx64 "] has been realigned\n"
+                      "to the %ld KiB page size: "
+                      "[0x%" PRIx64 ", 0x%" PRIx64 "]\n",
```
Comment (scottj97) [inline spike_main/spike.cc]: These changes are beneficial but independent of the new functionality you've added, so should be in its own commit.

File: spike_main/spike.cc [hunk 5]
```diff
@@ -176,20 +177,32 @@ static std::vector<mem_cfg_t> parse_mem_layout(const char* arg)
...
       exit(EXIT_FAILURE);
     }
 
-    res.push_back(mem_cfg_t(reg_t(base), reg_t(size)));
+    mem_cfg_t mem_region(base, size);
+    const uint64_t max_allowed_pa = (1ull << MAX_PADDR_BITS) - 1ull;
+    if ((mem_region.base + mem_region.size - 1) > max_allowed_pa) {
+      fprintf(stderr, "Unsupported memory region "
+                      "{base = 0x%" PRIx64 ", size = 0x%" PRIx64 "} specified\n"
+                      "    The largets accessible physical address "
```
Comment (scottj97) [inline spike_main/spike.cc]: "Largest" is misspelled

**Review by scottj97 — CHANGES_REQUESTED** [review]

File: spike_main/spike.cc (hunk 4 above)
Comment (aap-sc) [reply spike_main/spike.cc]: ok, will move this part to a separate PR then

Comment (scottj97) [reply spike_main/spike.cc]: Not a separate PR. It's related to the other change, so it's reasonable to include here in this PR.

File: spike_main/spike.cc (hunk 5 above)
Comment (aap-sc) [reply spike_main/spike.cc]: addressed

File: spike_main/spike.cc (hunk 4 above)
Comment (aap-sc) [reply spike_main/spike.cc]: The patch was refactored (a little bit)

**Comment by aap-sc** [comment]
To improve portability I've decided to stick with `unsigned long long` types and use PRIx64 only when it's required.

File: spike_main/spike.cc [hunk 6]
```diff
@@ -188,7 +190,20 @@ static std::vector<mem_cfg_t> parse_mem_layout(const char* arg)
...
 
-    res.push_back(mem_cfg_t(reg_t(base), reg_t(size)));
+    const unsigned long long max_allowed_pa = (1ull << MAX_PADDR_BITS) - 1ull;
+    assert(max_allowed_pa <= std::numeric_limits<reg_t>::max());
+    mem_cfg_t mem_region(base, size);
+    if ((mem_region.base + mem_region.size - 1) > max_allowed_pa) {
+      fprintf(stderr, "Unsupported memory region "
+                      "{base = 0x%" PRIx64 ", size = 0x%" PRIx64 "} specified\n"
+                      "    The largest accessible physical address "
+                      "is 0x%llX (defined by MAX_PADDR_BITS constant)\n",
```
Comment (scottj97) [inline spike_main/spike.cc]: Be consistent in usage of lower vs uppercase hex digits. This prints:
```
Unsupported memory region {base = 0xffffffffffff0000, size = 0x4000} specified
    The largest accessible physical address is 0xFFFFFFFFFFFFFF (defined by MAX_PADDR_BITS constant)
```

I also wonder if we could make the error message a little easier to read, instead of the user having to count all those digits? Such as:
```
Unsupported memory region {base = 0xffffffffffff0000, size = 0x4000} specified, which requires 64 paddr bits
    The largest accessible physical address is 0xFFFFFFFFFFFFFF (defined by MAX_PADDR_BITS constant, which is 56)
```

**Review by scottj97 — CHANGES_REQUESTED** [review]
A few nits and I think this is good.

"Redundant" is misspelled in the commit message for d2e1173f7ce1773aca1d3d3bb32b7d07bbd1938b

File: spike_main/spike.cc (hunk 6 above)
Comment (aap-sc) [reply spike_main/spike.cc]: > "Redundant" is misspelled

My bad. fixed.

> Be consistent in usage of lower vs uppercase hex digits.

Done.

> I also wonder if we could make the error message a little easier to read

Sure, however I'd like to use `__builtin_clzll` to calculate the number of required physical bits. Do you think it's appropriate? We already use `__builtin_expect` in likely/unlikely macro, so probably this should be ok.. Still I want to double-check

Comment (scottj97) [reply spike_main/spike.cc]: I'm okay with that, since it works in both GCC and clang. @aswaterman ?

Comment (aswaterman) [reply spike_main/spike.cc]: I've been trying to avoid that kind of thing.  Note that we only use `__builtin_expect` if `__GNUC__` is defined.

Please use this one: https://github.com/riscv-software-src/riscv-isa-sim/blob/ab3225a3ff687fda8b4180f9e4e0949a400d1247/riscv/arith.h#L174

Comment (aap-sc) [reply spike_main/spike.cc]: @aswaterman , 
> Please use this one:

Thanks!

> I also wonder if we could make the error message a little easier to read, instead of the user having to count all those digits? 

Addressed

File: spike_main/spike.cc [hunk 7]
```diff
@@ -176,20 +179,36 @@ static std::vector<mem_cfg_t> parse_mem_layout(const char* arg)
...
-    res.push_back(mem_cfg_t(reg_t(base), reg_t(size)));
+    const unsigned long long max_allowed_pa = (1ull << MAX_PADDR_BITS) - 1ull;
+    assert(max_allowed_pa <= std::numeric_limits<reg_t>::max());
+    mem_cfg_t mem_region(base, size);
+    auto last_pa_region = mem_region.base + mem_region.size - 1;
+    if (last_pa_region > max_allowed_pa) {
+      int bits_required = 64 - clz(last_pa_region);
+      fprintf(stderr, "Unsupported memory region "
+                      "{base = 0x%" PRIX64 ", size = 0x%" PRIX64 "} specified,"
+                      " which requires %d paddr bits\n"
```
Comment (aswaterman) [inline spike_main/spike.cc]: I suggest spelling out “physical address” on this line rather than abbreviating it “paddr”.

I’m happy with this PR otherwise. Thank you!

**Review by scottj97 — APPROVED** [review]
Excellent, thank you for your patience.

**Review by aswaterman — APPROVED** [review]
